File: src/decoders/attachment.ts
import { AES } from "~/api/private/aes";
import { aesKeys } from "~/api/private/keys";
import { AttachmentKind, type SessionHandle } from "~/models";
import type { Attachment } from "~/models/attachment";

export const decodeAttachment = (attachment: any, session: SessionHandle, parameters = {}): Attachment => {
  const name = attachment.L ?? "";
  const kind = attachment.G;
  const id = attachment.N;
  let url: string;

  if (kind === AttachmentKind.Link) {
    url = attachment.url ?? name;
  }
  else {
    const { iv, key } = aesKeys(session);

    // Create an entity of the attachment.
    const data = JSON.stringify({
      N: id,
      Actif: true,

      // Depending on the case, this could be customized.
      ...parameters
    });

    const encrypted = AES.encrypt(data, key, iv);
    url = `${session.information.url}/FichiersExternes/${encrypted}/${encodeURIComponent(name)}?Session=${session.information.id}`;
  }

  return {
    kind,
    name,
    url,
    id
  };
};

File: library/src/jvmMain/kotlin/ink/literate/pawnote/decoders/attachment.kt
package ink.literate.pawnote.decoders

import ink.literate.pawnote.api.private.AES
import ink.literate.pawnote.api.private.aesKeys
import ink.literate.pawnote.models.Attachment
import ink.literate.pawnote.models.AttachmentKind
import ink.literate.pawnote.models.SessionInformation

import java.net.URLEncoder
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*

actual fun decodeAttachment (attachment: JsonObject, sessionInfo: SessionInformation, parameters: JsonObject): Attachment {
    val name = attachment["L"]?.jsonPrimitive?.content ?: ""
    val kind = attachment["G"]!!.jsonPrimitive.int
    val id = attachment["N"]!!.jsonPrimitive.content
    val url: String

    if (kind == AttachmentKind.Link.code)
        url = attachment["url"]?.jsonPrimitive?.content ?: name
    else {
        val keys = aesKeys(sessionInfo)

        val data = JsonObject(parameters + buildJsonObject {
            put("N", id)
            put("Actif", true)
        })

        val encrypted = AES.encrypt(Json.encodeToString(data).toByteArray(), keys.key, keys.iv)
        url = "${sessionInfo.url}/FichiersExternes/${encrypted}/${URLEncoder.encode(name)}?Session=${sessionInfo.id}"
    }

    return Attachment(
        kind = AttachmentKind.fromInt(kind),
        name = name,
        url = url,
        id = id
    )
}